Fix pointer receiver handling in queryparams marshaler

ash2k opened the pull request:
**What this PR does / why we need it**:
`Time.MarshalQueryParameter()` and `Time.MarshalJSON()` try to handle nil pointer object (they call `t.IsZero()` which checks if t == nil) but fail because receiver is not a pointer so the dereference needed to pass it as receiver to these methods fails with npe.
In practice this happens with `Unstructured.SetDeletionTimestamp(Unstructured.GetDeletionTimestamp())`.

Here is the stacktrace of the failing test if receiver is not a pointer.
```go
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1407969]

goroutine 22 [running]:
testing.tRunner.func1(0xc4204f0680)
	/usr/local/Cellar/go/1.8/libexec/src/testing/testing.go:622 +0x29d
panic(0x1485a80, 0x1782bc0)
	/usr/local/Cellar/go/1.8/libexec/src/runtime/panic.go:489 +0x2cf
k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.(*Unstructured).SetDeletionTimestamp(0xc420030790, 0x0)
	/Users/ash2k/gopath/src/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:387 +0x29
k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.TestNilDeletionTimestamp(0xc4204f0680)
	/Users/ash2k/gopath/src/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go:40 +0x4c
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/43346)
<!-- Reviewable:end -->

smarterclayton commented:
@k8s-bot ok to test

Is this happening outside of a test case?

ash2k commented:
@smarterclayton Yes, but this is outside of Kubernetes. I'm trying to copy certain fields from one `Unstructured` to another one.
Wow, so many tests have failed - is it my fault?

ash2k commented:
I've removed the `MarshalJSON()` change - want to do it incrementally. I think I've fixed an issue where converter did not check if a type has a pointer receiver for the method when it is not a pointer-based type (i.e. when type is `Time` and it does not implement `Marshaller` interface but `*Time` does). See the diff.
One of the builds is failing however yesterday it was failing in a different way and I was not able to figure out what the issue was. And same now - I don't understand what is failing and why. Not familiar with the build system (yet). If someone could take a look...

/cc @smarterclayton

ash2k commented on staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go:
```diff
@@ -90,7 +90,14 @@ func customMarshalValue(value reflect.Value) (reflect.Value, bool) {
 	marshaler, ok := value.Interface().(Marshaler)
 	if !ok {
-		return reflect.Value{}, false
+		if !isPointerKind(value.Kind()) && value.CanAddr() {
```
I implemented this and then realized `json` package must be doing the same check. And it does https://github.com/golang/go/blob/9073af247d602dff4633710adf90c8b3c1869c45/src/encoding/json/encode.go#L392

ash2k replied:
No, the case where `Marshaler` interface is implemented by a type with a non-pointer receiver of the method is handled on line 91. Added a test - please correct me if I misunderstood you and this is not the test you meant.

ironcladlou replied:
@deads2k [like this](https://play.golang.org/p/kVXnIk2AIg)? If only the pointer type implements `Marshaler`, passing a value type to `customMarshalValue()` causes `CanAddr` to return false.

```
CanAddr reports whether the value's address can be obtained with Addr. Such
values are called addressable. A value is addressable if it is an element of a
slice, an element of an addressable array, a field of an addressable struct, or
the result of dereferencing a pointer. If CanAddr returns false, calling Addr
will panic.
```

ash2k commented:
I've just rebased onto master and it's all green. Could someone take a look? @kubernetes/sig-api-machinery-misc

ash2k commented:
39 days since I opened this PR. Is anybody reading my comments? :) I understand that this is not very important but it is clearly a bug and the fix seems to not be controversial.

ash2k commented:
/unassign @smarterclayton

ash2k commented:
/sig api-machinery

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/time_test.go:
```diff
@@ -171,3 +171,20 @@ func TestTimeProto(t *testing.T) {
+}
+
+func TestMarshalJSONNilPointer(t *testing.T) {
+	t.Skip("not sure if this should be supported")
```
@smarterclayton any reason this shouldn't be supported?

smarterclayton replied:
I'm not sure.  I'm somewhat concerned about this, but haven't had the time to think through implications.  I don't know that we have great coverage on bizarre time values.

ash2k replied:
I've removed the `MarshalJSON` test to reduce the scope of this PR. Hope it will help to get this merged faster...

deads2k commented on staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go:
```diff
@@ -90,7 +90,14 @@ func customMarshalValue(value reflect.Value) (reflect.Value, bool) {
 	marshaler, ok := value.Interface().(Marshaler)
 	if !ok {
-		return reflect.Value{}, false
+		if !isPointerKind(value.Kind()) && value.CanAddr() {
```
Previously, a call to `marshal(metav1.Now())` would marshal correctly.  Wouldn't the `CanAddr` fail now?  Worth adding a test since I'm sure someone else will have the question.

ironcladlou replied:
@ash2k I'm not sure we all came to a mutual understanding on this point. Do you guys agree that this is true:

>If only the pointer type implements Marshaler, passing a value type to customMarshalValue() causes CanAddr to return false.

And if so, do you agree that's a change from previous behavior, e.g.:

>Previously, a call to marshal(metav1.Now()) would marshal correctly. 

And if that's also true, is all that okay?

ash2k replied:
@ironcladlou @deads2k What `marshal()` are we talking about? 

`MarshalQueryParameter()` is a method on `metav1.Time`. It is only ever called via reflection from [`customMarshalValue()`](https://github.com/atlassian/kubernetes/blob/6ec536513a4c85af08f43def4bea1fdddde9f215/staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go#L85) which is only ever called from [`convertStruct()`](https://github.com/atlassian/kubernetes/blob/6ec536513a4c85af08f43def4bea1fdddde9f215/staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go#L162) which is only called for structs. So `metav1.Time` is always a field in a struct and it should always be possible to take the address of a field.

However I agree that if there was a method `marshal()` that tried to call `MarshalQueryParameter()` directly on a value (via reflection) and that value was `metav1.Time` then yes, it would be a change in behavior because it will not work (because of a pointer receiver). However I think this is a reasonable requirement to ask to pass a pointer to a value not just the value. That is how `encoding/json` package works: https://play.golang.org/p/jKkAXSJ9Ul

In the example above `MarshalJSON()` method is not picked up by `encoding/json` for the value/value case.

p.s. I was confused and now I understand that initial comment.

ash2k replied:
I'm actually happy to remove the change to the method receiver. The original issue has been fixed in #45167 anyway and I've audited all other places where `MarshalQueryParameter()` is called and they are not using a pointer `Time`.
I still believe the fix for reflection check should be merged though. WDYT?

ash2k replied:
^ done.

ironcladlou replied:
Thanks for the additional explanation... I agree that requiring a pointer argument is a reasonable requirement, and your detail about field addressability makes sense. I *think* everything makes sense now and looks good, and wish somebody else would chime in, but LGTM. :grin:

janetkuo commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go:
```diff
@@ -34,3 +34,8 @@ func TestUnstructuredList(t *testing.T) {
+
+func TestNilDeletionTimestamp(t *testing.T) {
+	var u Unstructured
+	u.SetDeletionTimestamp(u.GetDeletionTimestamp())
```
Check if the nested field is expected

janetkuo commented:
Can someone from api machinery review this too?

ash2k commented:
/kind bug

ash2k commented:
ptal @deads2k

lavalamp commented:
/assign @jpbetz

ironcladlou reviewed (commented):
I'm not sure how changing all these receivers to a pointer isn't resulting in panics in the test cases on the `IsZero()` calls. For example, here's a playground port of `TestMarshalQueryParameterNilPointer`: https://play.golang.org/p/ZN1xbpN34L

ash2k commented:
Here is why - the example you provided is missing the `IsZero()` method. This is a fixed example which works: https://play.golang.org/p/KpdauV9N5M

Explanation:
In your example `IsZero()` method is "inherited" from embedded `time.Time`. It looks like this:
```go
func (t Time) IsZero() bool {
	return t.sec == 0 && t.nsec == 0
}
```
Note that it has a value receiver hence when called on a pointer it will need to dereference it. Pointer is `nil` in your example hence the panic.

In `meta/v1/time.Time` this method is NOT "inherited" from embedded `time.Time` - it is provided explicitly and it looks like this:
```go
func (t *Time) IsZero() bool {
	if t == nil {
		return true
	}
	return t.Time.IsZero()
}
```
Note that it has a pointer receiver which checks if `t` can be dereferenced before doing so.

ironcladlou commented:
@ash2k Thanks for the explanation... between that, some more careful study of the language spec and some other articles I get it now. We're in the golang minefields...

ironcladlou commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go:
```diff
@@ -34,3 +34,17 @@ func TestUnstructuredList(t *testing.T) {
+	del := u.GetDeletionTimestamp()
+	if del != nil {
+		t.Errorf("unexpected non-nil deletion timestamp: %v", del)
+	}
```
Minor refactor for clarity?

```go
del := u.GetDeletionTimestamp()
if del != nil {
	t.Errorf("unexpected non-nil deletion timestamp: %v", del)
}
u.SetDeletionTimestamp(del)

ash2k replied:
I've added an extra check before `SetDeletionTimestamp()` is invoked if that is what you wanted? We need to check that getter returns `nil` after setter is invoked so I want to have both checks there.

ironcladlou reviewed (commented):
Just a minor aesthetic request

ash2k commented:
@ironcladlou PTAL

ash2k commented:
/retest

ash2k committed:
Pointer receiver support for MarshalQueryParameter()

ironcladlou commented:
/lgtm

Thanks for your patience with this one...

janetkuo commented:
/approve no-issue
/retest

ash2k commented:
/retest

janetkuo commented:
/retest

janetkuo commented:
@deads2k would you approve this?

ash2k commented:
/retest

dims commented:
/retest

ash2k commented:
/retest

ash2k commented:
Half a year later, this is still open :) Can someone take a look please? @ironcladlou

ironcladlou commented:
@ash2k, I already gave this a lgtm, but I don't think that's enough...

ash2k commented:
@deads2k PTAL.

p.s. Those builds failed because of something else and I cannot do `/retest`.

deads2k commented:
/retest
/approve no-issue
